File: app/src/main/java/nl/rsdt/japp/jotial/maps/kml/KmlFolder.java
package nl.rsdt.japp.jotial.maps.kml;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class KmlFolder {
    private Type type;
    private List<KmlPlaceMark> placeMarkList = new ArrayList<>();

    public KmlScoutingGroep toKmlOrganisatie() {
        if (type != Type.Organisatie) throw new AssertionError();
        if (placeMarkList.size() != 1) throw new AssertionError();
        return placeMarkList.get(0).toKmlScoutingGroep();
    }

    public List<KmlScoutingGroep> toKmlGroepen() {
        if (type != Type.Groepen) throw new AssertionError();
        List<KmlScoutingGroep> groepen = new LinkedList<>();
        for (KmlPlaceMark placeMark : placeMarkList){
            groepen.add(placeMark.toKmlScoutingGroep());
        }
        return groepen;
    }

    public List<KmlDeelgebied> toKmlDeelgebieden() {
        if (type != Type.Deelgebieden) throw new AssertionError();
        List<KmlDeelgebied> deelgebieden = new LinkedList<>();
        for (KmlPlaceMark placeMark : placeMarkList){
            deelgebieden.add(placeMark.toKmlDeelgebied());
        }
        return deelgebieden;
    }


    public enum Type {Organisatie, Groepen, Deelgebieden}

    public Type getType() {
        return type;
    }

    public void setType(Type type) {
        this.type = type;
    }
    public boolean addPlacemark(KmlPlaceMark placeMark){
        return placeMarkList.add(placeMark);
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/kml/KmlFolder.kt
package nl.rsdt.japp.jotial.maps.kml

import java.util.*

class KmlFolder {
    var type: Type? = null
    private val placeMarkList = ArrayList<KmlPlaceMark>()

    fun toKmlOrganisatie(): KmlScoutingGroep {
        if (type != Type.Organisatie) throw AssertionError()
        if (placeMarkList.size != 1) throw AssertionError()
        return placeMarkList[0].toKmlScoutingGroep()
    }

    fun toKmlGroepen(): List<KmlScoutingGroep> {
        if (type != Type.Groepen) throw AssertionError()
        val groepen = LinkedList<KmlScoutingGroep>()
        for (placeMark in placeMarkList) {
            groepen.add(placeMark.toKmlScoutingGroep())
        }
        return groepen
    }

    fun toKmlDeelgebieden(): List<KmlDeelgebied> {
        if (type != Type.Deelgebieden) throw AssertionError()
        val deelgebieden = LinkedList<KmlDeelgebied>()
        for (placeMark in placeMarkList) {
            deelgebieden.add(placeMark.toKmlDeelgebied())
        }
        return deelgebieden
    }


    enum class Type {
        Organisatie, Groepen, Deelgebieden, Unkown;


        companion object {

            fun parse(s: String): Type {
                when (s.toLowerCase()) {
                    "organisatie" -> return Organisatie
                    "groepen" -> return Groepen
                    "deelgebieden" -> return Deelgebieden
                    else -> return Unkown
                }

            }
        }
    }

    fun addPlacemark(placeMark: KmlPlaceMark): Boolean {
        return placeMarkList.add(placeMark)
    }
}
